File: Kakao/2019_매칭 점수.cpp
// 출제의도: "스트링패턴 매칭", C++은 string.find()를 활용. 링크점수 구할때 map 자료구조 활용시 유용. 입력크기가 작으므로 시간복잡도 고려하지 않고 구현만 하면 되는 문제.

#include <string>
#include <vector>
#include <unordered_map>
#include <algorithm>
#include <iostream>

using namespace std;

struct Page {
    string url;
    string body;
    vector<string> links;
};

string getUrl(string page) {
    int begin, end;
    int cur = page.find("<meta property=");
    cur = page.find("https://",cur);
    begin = cur;
    end = page.find("\"",cur);
    return page.substr(begin,end-begin);
}

string getBody(string page) {
    int begin, end;
    int cur = page.find("<body>");
    begin = cur+6;
    end = page.find("</body>",cur);
    return page.substr(begin,end-begin);
}

vector<string> getLinks(string body) {
    vector<string> links;
    int begin, end=0;
    while((begin=body.find("<a href=",end))!=-1) {
        begin = body.find("https://",begin);
        end = body.find("\"",begin);
        links.push_back(body.substr(begin,end-begin));
    }
    return links;
}

Page parse(string page) {
    Page pg;
    pg.url = getUrl(page);
    pg.body = getBody(page);
    pg.links = getLinks(pg.body);
    return pg;
}

int getBasicScore(string body, string word) {
    int basicScore=0;
    int begin, end=-1;
    transform(body.begin(),body.end(),body.begin(),::tolower);
    transform(word.begin(),word.end(),word.begin(),::tolower);
    while((begin=body.find(word,end+1))!=-1) {
        end = begin+word.size();
        if(begin-1>=0 && isalpha(body[begin-1]))   continue;
        if(end<body.size() && isalpha(body[end]))  continue;
        basicScore++;
    }
    return basicScore;
}

int solution(string word, vector<string> pages) {
    int answer = 0;
    double score_max=-1;
    unordered_map<string,double>   Score;  // url:score
    vector<Page>    pages_parsed;
    for(int i=0;i<pages.size();i++) {
        Page page = parse(pages[i]);
        pages_parsed.push_back(page);
    }
    for(int i=0;i<pages_parsed.size();i++) {
        int basicScore=getBasicScore(pages_parsed[i].body, word);
        Score[pages_parsed[i].url] += basicScore;
        for(int j=0;j<pages_parsed[i].links.size();j++) {
            Score[pages_parsed[i].links[j]] += (double)basicScore/pages_parsed[i].links.size();
        }
    }
    for(int i=0;i<pages_parsed.size();i++) {
        if(Score[pages_parsed[i].url]>score_max) {
            score_max = Score[pages_parsed[i].url];
            answer = i;
        }
    }

    return answer;
}

File: Kakao/2019_매칭 점수.swift
// 출제의도: "문자열처리 or 정규표현식", Time Complexity: 정규식 엔진 성능에 좌우됨, Space Complexity: O(n). (n: pages.size)
// 실질적인 출제의도는 html에 대한 이해, 정규 표현식 이라 판단됨. html에 대한 이해가 없으면 문제 이해가 오래 걸려서 시간 내에 풀 수 없음. url, 문자열 검색은 전형적인 정규표현식 이용 사례임.
// 정규 표현식에 대한 개념을 "손에 잡히는 10분 정규표현식"으로 정리. swift 언어는 C++보다 정규표현식 처리에 유리하다 판단함.
// 이 문제로 든 생각은, 카카오는 애초에 웹 개발자를 타깃을 두고 이들에게 유리한 문제를 출제를 한 것은 아닌가 함.
// 어쨌든, 이 문제를 통해 정규 표현식을 이용한 url, 문자열 검색을 경험해보았음.

import Foundation

extension String {
    func matches(pattern: String, option: NSRegularExpression.Options) -> [String] {
        do {
            let regex = try NSRegularExpression(pattern: pattern, options: option)
            let range = NSRange(self.startIndex..<self.endIndex, in: self)
            
            return regex.matches(in: self, options: [], range: range)
                .map { String(self[Range($0.range, in: self)!]) }
        } catch let error {
            print("regular expression pattern is invalid")
            return []
        }
    }
}

enum RegExType: String {
    case word = "[A-Z]+"
    case url = #"(?<=<meta property="og:url" content="https://)([^"]*)"#
    case link = #"(?<=<a href="https://)([^"]*)"#
}

func solution(_ word:String, _ pages:[String]) -> Int {
    var basicScore = [String: Double]()
    var linkScore = [String: Double]()
    
    pages.forEach {
        let url = $0.matches(pattern: RegExType.url.rawValue, option: .dotMatchesLineSeparators)
        let links = $0.matches(pattern: RegExType.link.rawValue, option: .dotMatchesLineSeparators)
        let score = $0.matches(pattern: RegExType.word.rawValue, option: .caseInsensitive)
            .filter({$0.lowercased() == word.lowercased()})
            .count

        basicScore[url[0]] = Double(score)
        links.forEach {
            linkScore[$0] = (linkScore[$0] ?? 0) + Double(score)/Double(links.count)
        }
    }
    
    var maximumMatchingScore: Double = -1
    return pages.enumerated().reduce(into: Int()) { ans, page in
        let url = page.element.matches(pattern: RegExType.url.rawValue, option: .dotMatchesLineSeparators)
        let matchingScore = (basicScore[url[0]] ?? 0)+(linkScore[url[0]] ?? 0)

        if matchingScore > maximumMatchingScore {
            maximumMatchingScore = matchingScore
            ans = page.offset
        }
    }
}
